File: src/main/java/leetcode/_13/RomanToInteger.java
package leetcode._13;

import java.util.Stack;

public class RomanToInteger {

    public int romanToInt(String s) {
        var sum = 0;
        var chars = s.toCharArray();
        var stack = new Stack<Integer>();

        for (var i = 0; i < chars.length; i++) {
            stack.push(valueOf(chars[i]));

            while ((i + 1) < chars.length && stack.peek() < valueOf(chars[i + 1])) {
                stack.push(valueOf(chars[++i]));
            }

            var curr = stack.pop();
            while (!stack.isEmpty()) {
                curr -= stack.pop();
            }

            sum += curr;
        }

        return sum;
    }

    private int valueOf(char c) {
        return switch (c) {
            case 'I' -> 1;
            case 'V' -> 5;
            case 'X' -> 10;
            case 'L' -> 50;
            case 'C' -> 100;
            case 'D' -> 500;
            case 'M' -> 1000;
            default -> 0;
        };
    }
}

File: src/main/java/leetcode/_13/roman_to_integer.go
package main

import "fmt"

var alphabet = map[byte]int{
	'I': 1,
	'V': 5,
	'X': 10,
	'L': 50,
	'C': 100,
	'D': 500,
	'M': 1000,
}

func romanToInt(s string) int {
	stack := make([]int, 10)
	N := 0
	sum := 0

	for i := 0; i < len(s); i++ {
		stack[N] = alphabet[s[i]] // stack push
		N++

		for (i+1) < len(s) && stack[N-1] < alphabet[s[i+1]] { // peek stack top -> (stack[N-1])
			i++
			stack[N] = alphabet[s[i]]
			N++
		}

		N--
		curr := stack[N] // stack pop

		for N > 0 {
			N--
			curr -= stack[N]
		}

		sum += curr
	}

	return sum
}

// ***************** Testing *****************
func main() {
	s := "MCMXCIV"
	fmt.Println(romanToInt(s)) // rez: 1994
}
